californium-core/src/test/java/org/eclipse/californium/core/test/MessageExchangeStoreTool.java

```diff
@@ -175,5 +186,65 @@ public CoapUdpTestStack getStack() {
 		public boolean isEmpty() {
 			return exchangeStore.isEmpty() && (stack == null || stack.isEmpty());
 		}
+
+		@Override
+		public void sendRequest(Request request) {
+			requestChecker.registerRequest(request);
+			super.sendRequest(request);
+		}
+
+		public RequestEventChecker getRequestChecker() {
+			return requestChecker;
+		}
+	}
+
+	public static class RequestEventChecker {
+
+		private Collection<Request> requests = Collections.synchronizedSet(new HashSet<Request>());
+
+		public void registerRequest(final Request request) {
+			requests.add(request);
+
+			request.addMessageObserver(new MessageObserverAdapter() {
+
+				@Override
+				public void onCancel() {
+					requests.remove(request);
+					System.out.println(request + " cancelled");
```

Comment:
I know, that the test frequently use `System.out` or `err`. should we keep this or change it to the logger?

Reply:
Oups sry, I forgot to remove this before to push :/